tests/unit/test_adapter.py

```diff
@@ -557,70 +559,159 @@ def test_main(
+            cba_config,
+            cba_log,
+            csp_config
+        )
+    assert err_msg in str(e)
+    assert err_msg in caplog.text
+
+
+@mock.patch('csp_billing_adapter.adapter.time.sleep')
+def test_metering_test_meter_billing_failure(
     mock_sleep,
-    mock_rand,
     cba_pm,
     cba_config,
     cba_log,
     caplog
 ):
+    err_exc = Exception('Mock Failure')
+    err_msg = (
+        "Fatal error while validating metering API access: "
+        f"{str(err_exc)}"
+    )
 
-    mock_get_pm.return_value = cba_pm
-    mock_setup_logging.return_value = cba_log
-    mock_get_config.return_value = cba_config
-    mock_rand.return_value = 0
+    cache, csp_config = initial_adapter_setup(
+        cba_pm.hook,
+        cba_config,
+        cba_log
+    )
+    assert cache != {}
+    assert csp_config != {}
 
-    # test catching CSP Billing Adapter exception caused by AttributeError
-    with pytest.raises(SystemExit) as e:
-        cba_main()
-    assert e.value.code == 2
-    assert (
-        "Billing adapter config is invalid. 'Config' object has no "
-        "attribute 'usage_metrics'"
-    ) in caplog.text
+    with mock.patch.object(
+        cba_pm.hook,
+        'meter_billing',
+        side_effect=err_exc
+    ):
+        # test catching CSPBillingAdapterException caused by
+        # meter_billing() fatal exception.
+        with pytest.raises(CSPBillingAdapterException) as e:
+            metering_test(
+                cba_pm.hook,
+                cba_config,
+                cba_log,
+                csp_config
+            )
+        assert err_msg in str(e)
+        assert err_msg in caplog.text
 
 
-@mock.patch('csp_billing_adapter.local_csp.randrange')
 @mock.patch('csp_billing_adapter.adapter.time.sleep')
-@mock.patch('csp_billing_adapter.adapter.get_plugin_manager')
-@mock.patch('csp_billing_adapter.adapter.get_config')
-@mock.patch('csp_billing_adapter.adapter.setup_logging')
-@pytest.mark.config('config_no_dimensions.yaml')
-def test_main_key_error_handling(
-    mock_setup_logging,
-    mock_get_config,
-    mock_get_pm,
+def test_metering_test_meter_billing_and_update_csp_config_failure(
     mock_sleep,
-    mock_rand,
     cba_pm,
     cba_config,
     cba_log,
     caplog
 ):
+    err_exc = Exception('Mock Failure')
+    err_msg1 = (
+        "Fatal error while validating metering API access: "
+        f"{str(err_exc)}"
+    )
+    err_msg2 = (
+        "Failed to save csp_config to datastore: "
+        f"{str(err_exc)}"
+    )
 
-    mock_get_pm.return_value = cba_pm
-    mock_setup_logging.return_value = cba_log
-    mock_get_config.return_value = cba_config
-    mock_rand.return_value = 0
+    cache, csp_config = initial_adapter_setup(
+        cba_pm.hook,
+        cba_config,
+        cba_log
+    )
+    assert cache != {}
+    assert csp_config != {}
 
-    # test catching CSP Billing Adapter exception caused by KeyError
-    with pytest.raises(SystemExit) as e:
-        cba_main()
```

Comment:
I think if you do not call main, you miss one line in the pytest run. Not sure if you care but wanted to make you aware
```
 pytest --cov=csp_billing_adapter
====================================================================== test session starts ======================================================================
platform linux -- Python 3.6.15, pytest-7.0.1, pluggy-1.0.0
rootdir: /home/kberger/repos/github/csp/csp-billing-adapter, configfile: setup.cfg, testpaths: tests
plugins: cov-4.0.0
collected 68 items

tests/unit/test_adapter.py .................                                                                                                              [ 25%]
tests/unit/test_bill_utils.py ...........                                                                                                                 [ 41%]
tests/unit/test_config.py ...                                                                                                                             [ 45%]
tests/unit/test_csp_cache.py .....                                                                                                                        [ 52%]
tests/unit/test_csp_config.py .                                                                                                                           [ 54%]
tests/unit/test_local_csp.py ....                                                                                                                         [ 60%]
tests/unit/test_memory_cache.py ....                                                                                                                      [ 66%]
tests/unit/test_memory_csp_config.py ....                                                                                                                 [ 72%]
tests/unit/test_product_api.py .                                                                                                                          [ 73%]
tests/unit/test_utils.py ..................                                                                                                               [100%]

---------- coverage: platform linux, python 3.6.15-final-0 -----------
Name                                       Stmts   Miss  Cover   Missing
------------------------------------------------------------------------
csp_billing_adapter/__init__.py                7      0   100%
csp_billing_adapter/adapter.py               148      1    99%   421
csp_billing_adapter/bill_utils.py            105      1    99%   260
csp_billing_adapter/config.py                 39      0   100%
csp_billing_adapter/csp_cache.py              30      0   100%
csp_billing_adapter/csp_config.py             11      0   100%
csp_billing_adapter/csp_hookspecs.py          13      0   100%
csp_billing_adapter/exceptions.py             12      0   100%
csp_billing_adapter/hookspecs.py              10      0   100%
csp_billing_adapter/local_csp.py              22      0   100%
csp_billing_adapter/memory_cache.py           21      0   100%
csp_billing_adapter/memory_csp_config.py      21      0   100%
csp_billing_adapter/product_api.py            13      0   100%
csp_billing_adapter/storage_hookspecs.py      16      0   100%
csp_billing_adapter/utils.py                  83      0   100%
------------------------------------------------------------------------
TOTAL                                        551      2    99%

Required test coverage of 90.0% reached. Total coverage: 99.64%

====================================================================== 68 passed in 2.15s ======================
```

Reply:
All of the statements of main() are still called as part of the test_main() unit test case in the unit tests.

What has changed is that before, when I wanted to test the various exception cases associated with the metering testing in the main() routine,m I needed to set up the appropriate mocks for all of the stuff that main() did that I didn't want to test.

But now that the metering test has moved to a separate routine, I can test those exception cases more directly.

However, the main() routine is still tested.

The two statements in entire code base that are currently not being covered by the unit testing are:
* the call of main() from the if __name__ == '__main__' check in adapter.py
* the pass statement in the else of the if comsumption_reporting == 'volume' check in bill_utils.py.